src/omero_zarr/masks.py

```diff
@@ -23,53 +39,199 @@ def image_masks_to_zarr(image, args):
 
         if len(mask_shapes) > 0:
             masks[roi.id.val] = mask_shapes
+            shape_count += len(mask_shapes)
 
-    print(f"Found {len(masks)} masks")
+    print(f"Found {shape_count} mask shapes in {len(masks)} ROIs")
 
-    if masks:
-        stack = masks_to_zarr(masks, image)
-        name = f"{image.id}_masks.zarr"
-        root = zarr.open_group(name, mode="w")
-        za = root.create(
-            "0",
-            shape=stack.shape,
-            chunks=(1, 1, size_y, size_x),
-            dtype=stack.dtype,
-        )
+    dtype = MASK_DTYPE_SIZE[int(args.mask_bits)]
 
-        za[:, :, :, :] = stack
-
-        print("Created", name)
+    if masks:
+        if args.group_rois:
+            for (roi_id, roi) in masks.items():
+                _save_masks([roi], image, str(roi_id), dtype)
+        else:
+            _save_masks(masks.values(), image, "0", dtype)
     else:
         print("No masks found on Image")
 
 
-def masks_to_zarr(masks, image):
-
-    # Create np nd-array same size as image
+def _save_masks(masks, image, roi_name, dtype):
     size_t = image.getSizeT()
+    size_c = image.getSizeC()
     size_z = image.getSizeZ()
-    size_x = image.getSizeX()
     size_y = image.getSizeY()
+    size_x = image.getSizeX()
+    image_shape = (size_t, size_c, size_z, size_y, size_x)
+
+    # Figure out whether we can flatten some dimensions
+    unique_dims = {
+        "T": set(),
+        "C": set(),
+        "Z": set(),
+    }
+    for shapes in masks:
+        for mask in shapes:
+            unique_dims["T"].add(unwrap(mask.theT))
+            unique_dims["C"].add(unwrap(mask.theC))
+            unique_dims["Z"].add(unwrap(mask.theZ))
+    ignored_dimensions = set()
+    print(unique_dims)
+    for d in "TCZ":
+        if unique_dims[d] == {None}:
+            ignored_dimensions.add(d)
+
+    name = f"{image.id}.zarr"
+    root = zarr.open(name)
+    if "masks" in root.group_keys():
+        out_masks = root.masks
+    else:
+        out_masks = root.create_group("masks")
+
+    mask_shape = list(image_shape)
+    for d in ignored_dimensions:
+        mask_shape[DIMENSION_ORDER[d]] = 1
+    print("Ignoring dimensions {}".format(ignored_dimensions))
+
+    za = out_masks.create_dataset(
+        roi_name,
+        shape=mask_shape,
+        chunks=(1, 1, 1, size_y, size_x),
+        dtype=dtype,
+        overwrite=True,
+    )
+    masks_to_labels(
+        masks, mask_shape, ignored_dimensions, check_overlaps=True, labels=za,
+    )
+
+    # Setting za.attrs[] doesn't work, so go via parent
+    if "0" in root:
+        image_name = "0"
+    else:
+        image_name = "omero://{}.zarr".format(image.id)
+    out_masks[roi_name].attrs["image"] = {
+        "array": image_name,
```

Comment:
This path might need to be relative, no? In the output it looks like it's talking about itself `0`.